### PR TITLE
Add docs for custom reason APIs

### PR DESCRIPTION
**Pull request by jjohannes**
This adds documentation for the _custom reason_ features added as part of #3814.

### CODE REVIEW

File: subprojects/docs/src/docs/release/notes.md [hunk 1]
```diff
@@ -148,6 +148,17 @@ This avoids a 2nd request for the JAR file when the POM is missing, making depen
 
 You can also use this to programmatically opt-in to using the new Gradle metadata format for one repository using `metadataSources { gradleMetadata() }`.
 
+### Specifying reasons for dependency declaration and resolution rules
+
+In complex builds, dependency resolution results can be hard to interpret. Sometimes the reason why a [dependency declaration](userguide/customizing_dependency_resolution_behavior.html) or a [rule](userguide/inspecting_dependencies.html#sec:dependency_declaration_reasons) was added to the build script can get lost. To improve on this situation, we extended all the corresponding APIs with the capability to define a _reason_ for each declaration or rule. These reasons are shown in dependency insight reports and error messages if the corrsponding declaration or rule influenced the resolution result. In the future, they will also be shown in build scans.
```
Comment (ljacomet) [inline subprojects/docs/src/docs/release/notes.md]: typo in _corresponding_ (missing the 'e')

File: subprojects/docs/src/docs/userguide/customizingDependencyResolutionBehavior.adoc [hunk 2]
```diff
@@ -17,7 +17,7 @@
 
 In most cases, Gradle's default dependency management will resolve the dependencies that you want in your build. In some cases, however, it can be necessary to tweak dependency resolution to ensure that your build receives exactly the right dependencies.
 
-There are a number of ways that you can influence how Gradle resolves dependencies.
+There are a number of ways that you can influence how Gradle resolves dependencies. All these mechanism offer an API to define a reason for why they are used. It is recommended to make use of this to make dependency resolution results more understandable. If any customization influenced the resolution result, the provided reason will show up in <<sec:identifying_reason_dependency_selection,dependency insight>> reports.
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/customizingDependencyResolutionBehavior.adoc]: > All these mechanism

Plural so 'mechanisms'

File: subprojects/docs/src/docs/userguide/inspectingDependencies.adoc [hunk 3]
```diff
@@ -84,3 +84,20 @@ Every Gradle project provides the task `dependencyInsight` to render the so-call
     <output args="-q dependencyInsight --dependency commons-codec --configuration scm"/>
 </sample>
 ++++
+
+[[sec:dependency_declaration_reasons]]
+=== Justifying dependency declarations with custom reasons
+
+When you declare a <<declaring_dependencies,dependency>> dependency or a <<sec:dependency_constraints,dependency constraint>>, you can provide a custom reason for the declaration. This makes your build script and the dependency insight report easier to comprehend.
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/inspectingDependencies.adoc]: the 'dependency' outside of the link needs to be removed

File: subprojects/docs/src/docs/userguide/inspectingDependencies.adoc [hunk 4]
```diff
@@ -84,3 +84,20 @@ Every Gradle project provides the task `dependencyInsight` to render the so-call
...
+
+++++
+<sample id="dependencyReason" dir="userguide/dependencyManagement/inspectingDependencies/dependencyReason" title="Giving a reason why a certain modules version is used in a dependency declaration">
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/inspectingDependencies.adoc]: > a certain modules version

Does not read well, either the 's' from 'modules' needs to go or this has to be rephrased.

Comment (jjohannes) [reply subprojects/docs/src/docs/userguide/inspectingDependencies.adoc]: removed the 's' and rephrased

File: subprojects/docs/src/docs/userguide/inspectingDependencies.adoc (hunk 3 above)
Comment (jjohannes) [reply subprojects/docs/src/docs/userguide/inspectingDependencies.adoc]: fixed

File: subprojects/docs/src/docs/userguide/customizingDependencyResolutionBehavior.adoc (hunk 2 above)
Comment (jjohannes) [reply subprojects/docs/src/docs/userguide/customizingDependencyResolutionBehavior.adoc]: fixed

**Comment by jjohannes** [comment]
Thanks @ljacomet. I fixed the typos.

File: subprojects/docs/src/docs/release/notes.md (hunk 1 above)
Comment (jjohannes) [reply subprojects/docs/src/docs/release/notes.md]: fixed

**Comment by jjohannes** [comment]
@bmuschko I merged this because this should go into 4.6. But if you have any additional comments I am happy to address them.

File: subprojects/docs/src/docs/release/notes.md [hunk 5]
```diff
@@ -148,6 +148,17 @@ This avoids a 2nd request for the JAR file when the POM is missing, making depen
 
 You can also use this to programmatically opt-in to using the new Gradle metadata format for one repository using `metadataSources { gradleMetadata() }`.
 
+### Specifying reasons for dependency declaration and resolution rules
+
+In complex builds, dependency resolution results can be hard to interpret. Sometimes the reason why a [dependency declaration](userguide/customizing_dependency_resolution_behavior.html) or a [rule](userguide/inspecting_dependencies.html#sec:dependency_declaration_reasons) was added to the build script can get lost. To improve on this situation, we extended all the corresponding APIs with the capability to define a _reason_ for each declaration or rule. These reasons are shown in dependency insight reports and error messages if the corresponding declaration or rule influenced the resolution result. In the future, they will also be shown in build scans.
```
Comment (bmuschko) [inline subprojects/docs/src/docs/release/notes.md]: I am not quite sure how the reason can get lost. Maybe users don't know how to look for it via the dependencyInsight report or a build scan? We should be more specific here.

File: subprojects/docs/src/docs/release/notes.md [hunk 6]
```diff
@@ -148,6 +148,17 @@ This avoids a 2nd request for the JAR file when the POM is missing, making depen
 
 You can also use this to programmatically opt-in to using the new Gradle metadata format for one repository using `metadataSources { gradleMetadata() }`.
 
+### Specifying reasons for dependency declaration and resolution rules
```
Comment (bmuschko) [inline subprojects/docs/src/docs/release/notes.md]: I'd try to indicate that the "reason" is provided by the user e.g. by saying "Ability to specify custom reasons for dependency declaration and resolution rules".

File: subprojects/docs/src/docs/userguide/customizingDependencyResolutionBehavior.adoc [hunk 7]
```diff
@@ -17,7 +17,7 @@
 
 In most cases, Gradle's default dependency management will resolve the dependencies that you want in your build. In some cases, however, it can be necessary to tweak dependency resolution to ensure that your build receives exactly the right dependencies.
 
-There are a number of ways that you can influence how Gradle resolves dependencies.
+There are a number of ways that you can influence how Gradle resolves dependencies. All these mechanisms offer an API to define a reason for why they are used. It is recommended to make use of this to make dependency resolution results more understandable. If any customization influenced the resolution result, the provided reason will show up in <<sec:identifying_reason_dependency_selection,dependency insight>> reports.
```
Comment (bmuschko) [inline subprojects/docs/src/docs/userguide/customizingDependencyResolutionBehavior.adoc]: All _of_ these mechanisms...

Comment (bmuschko) [inline subprojects/docs/src/docs/userguide/customizingDependencyResolutionBehavior.adoc]: "It is recommended to make use of this to make..." sounds somewhat clunky. We should reword.

File: subprojects/docs/src/docs/userguide/inspectingDependencies.adoc [hunk 8]
```diff
@@ -84,3 +84,20 @@ Every Gradle project provides the task `dependencyInsight` to render the so-call
     <output args="-q dependencyInsight --dependency commons-codec --configuration scm"/>
 </sample>
 ++++
+
+[[sec:dependency_declaration_reasons]]
+=== Justifying dependency declarations with custom reasons
+
+When you declare a <<declaring_dependencies,dependency>> or a <<sec:dependency_constraints,dependency constraint>>, you can provide a custom reason for the declaration. This makes your build script and the dependency insight report easier to comprehend.
```
Comment (bmuschko) [inline subprojects/docs/src/docs/userguide/inspectingDependencies.adoc]: Does it really make a build script easier to understand? I don't think so. The build script can still be a complete mess.

Comment (bmuschko) [inline subprojects/docs/src/docs/userguide/inspectingDependencies.adoc]: I'd also rather say "easier to interpret".

Comment (bmuschko) [inline subprojects/docs/src/docs/userguide/inspectingDependencies.adoc]: Potentially link to section that explains "dependency insight report". Is this reason also reflected in a build scan? If it is then we should add a screenshot similar to the other sections in this chapter.

File: subprojects/docs/src/samples/userguide/artifacts/componentModuleMetadata/build.gradle [hunk 9]
```diff
@@ -4,8 +4,8 @@ apply plugin: 'java'
 dependencies {
     modules {
         module("com.google.collections:google-collections") {
-            replacedBy("com.google.guava:guava")
+            replacedBy("com.google.guava:guava", "Guava replaced google-collections")
```
Comment (bmuschko) [inline subprojects/docs/src/samples/userguide/artifacts/componentModuleMetadata/build.gradle]: I'd like to see a better reason here. Maybe something like "Google decided to rename and relocate Google Collections". As a user I might also be interested in a link to additional resources. Maybe something you could discuss as future improvement.

File: subprojects/docs/src/docs/release/notes.md (hunk 5 above)
Comment (jjohannes) [reply subprojects/docs/src/docs/release/notes.md]: With "reason can get lost" I was not referring to the reason concept in Gradle but just the reason in the head of the author when editing the build script. I reworded it.

File: subprojects/docs/src/docs/release/notes.md (hunk 6 above)
Comment (jjohannes) [reply subprojects/docs/src/docs/release/notes.md]: Good suggested. I changed it.

File: subprojects/docs/src/docs/userguide/customizingDependencyResolutionBehavior.adoc (hunk 7 above)
Comment (jjohannes) [reply subprojects/docs/src/docs/userguide/customizingDependencyResolutionBehavior.adoc]: fixed

Comment (jjohannes) [reply subprojects/docs/src/docs/userguide/customizingDependencyResolutionBehavior.adoc]: done

File: subprojects/docs/src/docs/userguide/inspectingDependencies.adoc (hunk 8 above)
Comment (jjohannes) [reply subprojects/docs/src/docs/userguide/inspectingDependencies.adoc]: Ok not the completed build script maybe but that part. I thought that's somehow clear from the context. I clarified.

Comment (jjohannes) [reply subprojects/docs/src/docs/userguide/inspectingDependencies.adoc]: fixed

Comment (jjohannes) [reply subprojects/docs/src/docs/userguide/inspectingDependencies.adoc]: No this is not yet in build scans. But the team knows that we should update this part of the documentation once it is.

The _dependency insight report section_ is right above this one.

File: subprojects/docs/src/samples/userguide/artifacts/componentModuleMetadata/build.gradle (hunk 9 above)
Comment (jjohannes) [reply subprojects/docs/src/samples/userguide/artifacts/componentModuleMetadata/build.gradle]: I don't want to make the text too long in the example. That's why I kept it vague. I reworded it a bit but kept it short. It's not only a relocation, as Guava includes not only collections but also other libraries

In general you can add a longer reason (e.g., include a link).

**Comment by jjohannes** [comment]
Thanks @bmuschko for the feedback. I addressed it in https://github.com/gradle/gradle/commit/cf0f2df369e94fc717aadbda5d4d75b4198199cf.

File: subprojects/docs/src/docs/release/notes.md (hunk 5 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/release/notes.md]: That wasn't necessarily clear to me. Thanks!

**Pull request merged**